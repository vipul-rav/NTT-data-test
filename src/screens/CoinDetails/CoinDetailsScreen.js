import React, { memo, useEffect, useState } from 'react';
import { useParams, Link } from 'react-router-dom';
import Alert from 'react-bootstrap/Alert';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import { LoaderComponent } from '../../components/Loader';

import { FormLabelGroupComponent } from '../../components/FormLabelGroup';
import { API_BASE_URL } from '../../constant/stringConstant';

const CoinDetailsScreen = memo(() => {
  const [coinDetails, updateCoinDetails] = useState(null);
  //show loader- loader can handle with middleware and redux globally in App
  const [loading, updateLoading] = useState(true);
  //read coin id to get details
  const { id } = useParams();

  useEffect(() => {
    const getCoinDetails = async () => {
      try {
        const result = await fetch(`${API_BASE_URL}api/v3/coins/${id}?localization=false`);
        const response = await result.json();
        updateCoinDetails(response);
        updateLoading(false);
      } catch (e) {
        updateLoading(false);
      }
    };
    getCoinDetails();
  }, [id]);

  if (loading) {
    return <LoaderComponent />;
  }

  const { name, symbol, hashing_algorithm, description, market_data, links, genesis_date, error } = coinDetails;

  return (
    <Container>
      <Row>
        <Col>
          <Link to="/">Back</Link>
        </Col>
      </Row>
      <Row>
        <Col>
          <strong>Coin Details</strong>
        </Col>
      </Row>
      <Row>
        <Col>
          {error ? (
            <Alert variant="danger">{error}</Alert>
          ) : (
            <Form>
              <FormLabelGroupComponent label="Name" value={name} />
              <FormLabelGroupComponent label="Symbol" value={symbol} />
              <FormLabelGroupComponent label="Hashing Alogorithm" value={hashing_algorithm} />
              <FormLabelGroupComponent label="Description" value={description?.en} />
              <FormLabelGroupComponent label="Market Cap" value={market_data?.market_cap?.eur} />
              <FormLabelGroupComponent label="Homepage" value={links?.homepage} />
              <FormLabelGroupComponent label="Genesis Date" value={genesis_date} />
            </Form>
          )}
        </Col>
      </Row>
    </Container>
  );
});

export { CoinDetailsScreen };
