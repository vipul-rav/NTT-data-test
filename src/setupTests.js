import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

const AllTheProviders = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};
const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
