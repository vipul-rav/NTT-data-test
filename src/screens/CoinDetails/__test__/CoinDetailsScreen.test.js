import React from 'react';
import { waitForElementToBeRemoved, cleanup } from '@testing-library/react';
import { render } from '../../../setupTests';
import { CoinDetailsScreen } from '../CoinDetailsScreen';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    id: 'bitcoin'
  })
}));

const setup = (overridesProps) => {
  const props = {
    ...overridesProps
  };

  const { container, getByText } = render(<CoinDetailsScreen {...props} />);

  return {
    props,
    container,
    getByText
  };
};

describe('CoinDetailsScreen', () => {
  afterEach(cleanup);
  it('should render', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });
  it('should render- API call', async () => {
    const { container, getByText } = setup();

    await waitForElementToBeRemoved(() => getByText('Please Wait...'));

    expect(container).toMatchSnapshot();
  });
});
