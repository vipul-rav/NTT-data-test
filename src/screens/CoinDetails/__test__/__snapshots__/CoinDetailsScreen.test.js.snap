// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinDetailsScreen should render 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="spinner-border"
        />
        Please Wait...
      </div>
    </div>
  </div>
</div>
`;

exports[`CoinDetailsScreen should render- API call 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <a
          href="/"
        >
          Back
        </a>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <strong>
          Coin Details
        </strong>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <form
          class=""
        >
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Name
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                Bitcoin
              </label>
            </div>
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Symbol
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                btc
              </label>
            </div>
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Hashing Alogorithm
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                SHA-256
              </label>
            </div>
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Description
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                Bitcoin is the first successful internet money based on peer-to-peer technology; whereby no central bank or authority is involved in the transaction and production of the Bitcoin currency. It was created by an anonymous individual/group under the name, Satoshi Nakamoto. The source code is available publicly as an open source project, anybody can look at it and be part of the developmental process.

Bitcoin is changing the way we see money as we speak. The idea was to produce a means of exchange, independent of any central authority, that could be transferred electronically in a secure, verifiable and immutable way. It is a decentralized peer-to-peer internet currency making mobile payment easy, very low transaction fees, protects your identity, and it works anywhere all the time with no central authority and banks.

Bitcoin is designed to have only 21 million BTC ever created, thus making it a deflationary currency. Bitcoin uses the &lt;a href="https://www.coingecko.com/en?hashing_algorithm=SHA-256"&gt;SHA-256&lt;/a&gt; hashing algorithm with an average transaction confirmation time of 10 minutes. Miners today are mining Bitcoin using ASIC chip dedicated to only mining Bitcoin, and the hash rate has shot up to peta hashes.

Being the first successful online cryptography currency, Bitcoin has inspired other alternative currencies such as &lt;a href="https://www.coingecko.com/en/coins/litecoin"&gt;Litecoin&lt;/a&gt;, &lt;a href="https://www.coingecko.com/en/coins/peercoin"&gt;Peercoin&lt;/a&gt;, &lt;a href="https://www.coingecko.com/en/coins/primecoin"&gt;Primecoin&lt;/a&gt;, and so on.

The cryptocurrency then took off with the innovation of the turing-complete smart contract by &lt;a href="https://www.coingecko.com/en/coins/ethereum"&gt;Ethereum&lt;/a&gt; which led to the development of other amazing projects such as &lt;a href="https://www.coingecko.com/en/coins/eos"&gt;EOS&lt;/a&gt;, &lt;a href="https://www.coingecko.com/en/coins/tron"&gt;Tron&lt;/a&gt;, and even crypto-collectibles such as &lt;a href="https://www.coingecko.com/buzz/ethereum-still-king-dapps-cryptokitties-need-1-billion-on-eos"&gt;CryptoKitties&lt;/a&gt;.
              </label>
            </div>
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Market Cap
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                391014498369
              </label>
            </div>
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Homepage
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                http://www.bitcoin.org
                
                
              </label>
            </div>
          </div>
          <div
            class="row"
          >
            <label
              class="form-label col-form-label col-sm-2"
            >
              <storng>
                Genesis Date
              </storng>
            </label>
            <div
              class="col"
            >
              <label
                class="form-label col-form-label col-sm-10"
              >
                2009-01-03
              </label>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
